packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -45,47 +49,69 @@ library LibProposing {
         TaikoData.Config memory config,
         AddressResolver resolver,
         TaikoData.BlockMetadataInput memory input,
+        TaikoData.ProverAssignment memory assignment,
         bytes calldata txList
     )
         internal
         returns (TaikoData.BlockMetadata memory meta)
     {
-        {
-            address proposer = resolver.resolve("proposer", true);
-            if (proposer != address(0) && msg.sender != proposer) {
-                revert L1_PERMISSION_DENIED();
-            }
+        // Check proposer
+        address proposer = resolver.resolve("proposer", true);
+        if (proposer != address(0) && msg.sender != proposer) {
+            revert L1_INVALID_PROPOSER();
         }
-        // Try to select a prover first to revert as earlier as possible
-        (address assignedProver, uint32 rewardPerGas) = IProverPool(
-            resolver.resolve("prover_pool", false)
-        ).assignProver(state.slotB.numBlocks, state.slotC.feePerGas);
 
-        assert(assignedProver != address(1));
+        // Check prover assignment
+        if (
+            assignment.prover == address(0) || assignment.prover == address(1)
```

Comment:
Looks like right now a proposer **must** select a prover when proposing? so any proposer must know at least one prover before proposing?

Reply:
Yes, that's the idea